packages/core/snapshot/read.go

```diff
@@ -85,68 +78,129 @@ func readSnapshotHeader(reader io.ReadSeeker) (*ledger.SnapshotHeader, error) {
+func readOutputWithMetadata(reader io.ReadSeeker) (outputMetadatas []*ledger.OutputWithMetadata, err error) {
+	var outputsLen int64
+	if err := binary.Read(reader, binary.LittleEndian, &outputsLen); err != nil {
+		return nil, fmt.Errorf("unable to read outputsWithMetadata bytes len: %w", err)
+	}
 
-	if len(data) > 0 {
-		outputMetadatas = make([]*ledger.OutputWithMetadata, 0)
-		_, err = serix.DefaultAPI.Decode(context.Background(), data, &outputMetadatas, serix.WithValidation())
-		if err != nil {
-			return nil, err
-		}
+	outputsBytes := make([]byte, outputsLen)
+	if err := binary.Read(reader, binary.LittleEndian, outputsBytes); err != nil {
+		return nil, fmt.Errorf("unable to read outputsWithMetadata: %w", err)
+	}
 
-		for _, o := range outputMetadatas {
-			o.SetID(o.M.OutputID)
-			o.Output().SetID(o.M.OutputID)
-		}
+	outputMetadatas = make([]*ledger.OutputWithMetadata, 0)
+	_, err = serix.DefaultAPI.Decode(context.Background(), outputsBytes, &outputMetadatas, serix.WithValidation())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, o := range outputMetadatas {
+		o.SetID(o.M.OutputID)
+		o.Output().SetID(o.M.OutputID)
 	}
 
 	return
 }
 
 // readEpochDiffs consumes a map of EpochDiff from the given reader.
-func readEpochDiffs(scanner *bufio.Scanner) (epochDiffs map[epoch.Index]*ledger.EpochDiff, err error) {
-	epochDiffs = make(map[epoch.Index]*ledger.EpochDiff)
-
-	scanner.Scan()
-	data := scanner.Bytes()
-	if len(data) > 0 {
-		_, err = serix.DefaultAPI.Decode(context.Background(), data, &epochDiffs, serix.WithValidation())
+func readEpochDiffs(reader io.ReadSeeker) (epochDiffs *ledger.EpochDiff, err error) {
+	spent := make([]*ledger.OutputWithMetadata, 0)
+	created := make([]*ledger.OutputWithMetadata, 0)
+
+	// read spent
+	var spentLen int64
+	if err := binary.Read(reader, binary.LittleEndian, &spentLen); err != nil {
+		return nil, fmt.Errorf("unable to read epochDiffs spent len: %w", err)
+	}
+	spentLenInt := int(spentLen)
+	for i := 0; i < spentLenInt; {
+		s, err := readOutputWithMetadata(reader)
 		if err != nil {
-			return nil, errors.Errorf("failed to parse epochDiffs from bytes: %w", err)
+			return nil, fmt.Errorf("unable to read epochDiffs spent: %w", err)
 		}
+		spent = append(spent, s...)
+		i += len(s)
+	}
 
-		for _, epochdiff := range epochDiffs {
-			for _, spentOutput := range epochdiff.Spent() {
-				spentOutput.SetID(spentOutput.M.OutputID)
-				spentOutput.Output().SetID(spentOutput.M.OutputID)
-			}
-			for _, createdOutput := range epochdiff.Created() {
-				createdOutput.SetID(createdOutput.M.OutputID)
-				createdOutput.Output().SetID(createdOutput.M.OutputID)
-			}
+	// read created
+	var createdLen int64
+	if err := binary.Read(reader, binary.LittleEndian, &createdLen); err != nil {
+		return nil, fmt.Errorf("unable to read epochDiffs created len: %w", err)
+	}
+	createdLenInt := int(createdLen)
+	for i := 0; i < createdLenInt; {
+		c, err := readOutputWithMetadata(reader)
```

Comment:
Why the for loop if this method returns multiple outputs?

Reply:
OutputWithMetadatas are read/write in chunks, so every time it's read, it returns a slice.